packages/ifpack2/src/Ifpack2_Details_ChebyshevKernel_def.hpp

```diff
@@ -0,0 +1,511 @@
+int64_t
+chebyshev_kernel_vector_launch_parameters (int64_t numRows,
+                                                 int64_t nnz,
+                                                 int64_t rows_per_thread,
+                                                 int& team_size,
+                                                 int& vector_length)
+{
+  using execution_space = typename ExecutionSpace::execution_space;
+
+  int64_t rows_per_team;
+  int64_t nnz_per_row = nnz/numRows;
+
+  if (nnz_per_row < 1) {
+    nnz_per_row = 1;
+  }
+
+  if (vector_length < 1) {
+    vector_length = 1;
+    while (vector_length<32 && vector_length*6 < nnz_per_row) {
+      vector_length *= 2;
+    }
+  }
+
+  // Determine rows per thread
+  if (rows_per_thread < 1) {
+#ifdef KOKKOS_ENABLE_CUDA
+    if (std::is_same<Kokkos::Cuda, execution_space>::value) {
+      rows_per_thread = 1;
+    }
+    else
+#endif
+    {
+      if (nnz_per_row < 20 && nnz > 5000000) {
+        rows_per_thread = 256;
+      }
+      else {
+        rows_per_thread = 64;
+      }
+    }
+  }
+
+#ifdef KOKKOS_ENABLE_CUDA
+  if (team_size < 1) {
+    if (std::is_same<Kokkos::Cuda,execution_space>::value) {
+      team_size = 256/vector_length;
+    }
+    else {
+      team_size = 1;
+    }
+  }
+#endif
+
+  rows_per_team = rows_per_thread * team_size;
+
+  if (rows_per_team < 0) {
+    int64_t nnz_per_team = 4096;
+    int64_t conc = execution_space::concurrency ();
+    while ((conc * nnz_per_team * 4 > nnz) &&
+           (nnz_per_team > 256)) {
+      nnz_per_team /= 2;
+    }
+    rows_per_team = (nnz_per_team + nnz_per_row - 1) / nnz_per_row;
+  }
+
+  return rows_per_team;
+}
+
+// W := alpha * D * (B - A*X) + beta * W.
+template<class WVector,
+         class DVector,
+         class BVector,
+         class AMatrix,
+         class XVector,
+         class Scalar>
+static void
+chebyshev_kernel_vector
+(const Scalar& alpha,
+ const WVector& w,
+ const DVector& d,
+ const BVector& b,
+ const AMatrix& A,
+ const XVector& x,
+ const Scalar& beta)
+{
+  using execution_space = typename AMatrix::execution_space;
+
+  if (A.numRows () == 0) {
+    return;
+  }
+
+  int team_size = -1;
+  int vector_length = -1;
+  int64_t rows_per_thread = -1;
+
+  const int64_t rows_per_team =
+    chebyshev_kernel_vector_launch_parameters<execution_space>
+      (A.numRows (), A.nnz (), rows_per_thread, team_size, vector_length);
+  int64_t worksets = (b.extent (0) + rows_per_team - 1) / rows_per_team;
+
+  using Kokkos::Dynamic;
+  using Kokkos::Schedule;
+  using Kokkos::TeamPolicy;
+  using policy_type = TeamPolicy<execution_space, Schedule<Dynamic>>;
+  const char kernel_label[] = "chebyshev_kernel_vector";
+  policy_type policy (1, 1);
+  if (team_size < 0) {
+    policy = policy_type (worksets, Kokkos::AUTO, vector_length);
+  }
+  else {
+    policy = policy_type (worksets, team_size, vector_length);
+  }
+
+  // Canonicalize template arguments to avoid redundant instantiations.
```

Comment:
We should actually canonicalize them to their nonowning variants (`MemoryTraits<Unmanaged>`).